compiler/src/dotty/tools/dotc/typer/RefChecks.scala

```diff
@@ -926,6 +928,81 @@ object RefChecks {
   }
 
   val NoLevelInfo: RefChecks.OptLevelInfo = new OptLevelInfo()
+
+  /**
+   * Verify that references in the user-defined @implicitNotFound message are valid.
+   * (i.e. they refer to a type variable that really occurs in the signature of the annotated symbol.)
+   */
+  private object checkImplicitNotFoundAnnotation:
+
+    /**
+     * Warns if the class or trait has an @implicitNotFound annotation
+     * with invalid type variable references.
+     */
+    def template(sd: SymDenotation)(using Context): Unit =
+      for
+        annotation <- sd.annotations
+        if annotation.symbol == ctx.definitions.ImplicitNotFoundAnnot
+        (arg, l) <- annotation.arguments.collect {
+          case l@Literal(c: Constant) => (c.stringValue, l)
+        }
+      do forEachTypeVariableReferenceIn(arg) { case (ref, start) =>
+        if !sd.typeParams.exists(_.denot.name.show == ref) then
+          reportInvalidReferences(l, ref, start, sd)
+      }
+
+    /**
+     * Warns if the def has parameters with an @implicitNotFound annotation
+     * with invalid type variable references.
+     */
+    def defDef(sd: SymDenotation)(using Context): Unit =
+      for
+        param <- sd.paramSymss.flatten
+        annotation <- param.annotations
+        if annotation.symbol == ctx.definitions.ImplicitNotFoundAnnot
+        (arg, l) <- annotation.arguments.collect {
```

Comment:
`Annotation` has an `argumentConstant` method you should be able to use here.

Reply:
I am trying to apply that but the Constant type doesn't seem to have the source position associated with it (by which I mean a `.span`). This means I would have to do some workaround for accurate warnings. I will give it a try.

Reply:
Ah I see, yes types don't have positions indeed, only trees do, so you could maybe use `argument` instead and then get the constant out.

Reply:
Ah yes that worked nicely! I'll try resolve the merge conflict that has _emerged_ (...) and then push the changes!